### DISCUSSION TITLE
Usage for packing on iOS

### DISCUSSION THREAD

**Discussion by yongjie0203** [xjasonlyu/tun2socks#171] [Q&A]
`
class PacketTunnelProvider: NEPacketTunnelProvider {
  
    var lastPath:NWPath?
    var started:Bool = false
    var proxyPort: UInt16! = 8088
   
    
    override func startTunnel(options: [String : NSObject]?, completionHandler: @escaping (Error?) -> Void) {
        // Add code here to start the process of connecting the tunnel.
        
        NSLog("--------------------------tunnel---------------------")
      
        let ipv4Settings = NEIPv4Settings(addresses: ["192.169.89.1"], subnetMasks: ["255.255.255.0"])
        let networkSettings = NEPacketTunnelNetworkSettings(tunnelRemoteAddress: "8.8.8.8")
        networkSettings.mtu = 1500
        networkSettings.ipv4Settings = ipv4Settings
        setTunnelNetworkSettings(networkSettings) {
            error in
            guard error == nil else {
                completionHandler(error)
                return
            }
                completionHandler(nil)
        }

        let tunFd = getFD()
        let fdStr = "fd://\(String(tunFd!))"
        let key = EngineKey()
        key.mtu = 1500
        key.device = fdStr
        key.logLevel = "debug"
        key.proxy = "socks5://127.0.0.1:8088"
        EngineInsert(key)
        EngineStart()
        self.started = true
        
    }
    `

以上为代码，大佬帮忙看下哪里有问题，一直没数据进代理服务器地址

**Comment by ding2chen** [comment]
请问解决了吗？ios和mac 通过networkExtension 方式都有这个问题

**Reply by yongjie0203** [reply]
我按苹果推荐的方式做了调整  ，修改了tun2socks部分源码，目前流量可以被转发到socks5代理服务器，但我这边还有问题没有解决。

**Reply by chen2ding** [reply]
是采用苹果packetFlow 读写？不知道为啥直接fd不行

**Reply by yongjie0203** [reply]
对的 苹果官方不推荐使用文件描述符读写，目前看直接从文件描述符读写需要处理iOS额外在IP包前加的4个字节头。tun2socks iobase包的处理貌似包含了对这4个字节的处理。但我目前也不清楚为啥不能使用fd。。我目前遇到的问题是网络请求很多会超时，还不太确定是配置问题或者其他地方有问题。

**Reply by ding2chen** [reply]
![image](https://user-images.githubusercontent.com/108846118/208009077-c6e9f4d3-7dbe-4522-bfd3-4033c19ced37.png)
ios 应该会编译这个文件吧，这里没有处理tun网卡数据包的前四个字节呢

**Reply by yongjie0203** [reply]
可以自行改成4 试试

**Reply by yongjie0203** [reply]
记录下需要注意的：
1. 网卡信息配置，网卡配置信息有问题也会导致流量无法拦截。
2. swift到go 流量数据的传输方式数据类型影响。
